mirgecom/simutil.py

```diff
@@ -74,92 +88,273 @@ def inviscid_sim_timestep(discr, state, t, dt, cfl, eos,
+
+    Parameters
+    ----------
+    eos: mirgecom.eos.GasEOS
+        Implementing the pressure and temperature functions for
+        returning pressure and temperature as a function of the state *q*.
+    q
+        State array which expects at least the conserved quantities
+        (mass, energy, momentum) for the fluid at each point. For multi-component
+        fluids, the conserved quantities should include
+        (mass, energy, momentum, species_mass), where *species_mass* is a vector
+        of species masses.
+    freq: nstatus
+        An integer denoting the step frequency for performing status checks.
+    freq: nviz
+        An integer denoting the step frequency for writing vtk output.
+    """
+    exception = None
+    comm = discr.mpi_communicator
+    rank = 0
+    if comm is not None:
+        rank = comm.Get_rank()
+
+    try:
+        # Status checks
+        if check_step(step=step, interval=nstatus):
+            from mirgecom.fluid import split_conserved
+
+            #        if constant_cfl is False:
+            #            current_cfl = get_inviscid_cfl(discr=discr, q=q,
+            #                                           eos=eos, dt=dt)
+
+            cv = split_conserved(discr.dim, q)
+            dependent_vars = eos.dependent_vars(cv)
+            msg = make_status_message(discr=discr,
+                                    t=t, step=step, dt=dt, cfl=cfl,
+                                    dependent_vars=dependent_vars)
+            if rank == 0:
+                logger.info(msg)
+
+            # Check the health of the simulation
+            sim_healthcheck(discr, eos, q, step=step, t=t)
+
+            # Compare with exact solution, if provided
+            if exact_soln is not None:
+                compare_with_analytic_solution(
+                    discr, eos, q, exact_soln,
+                    step=step, t=t, exittol=exittol
+                )
+
+        # Visualization
+        if check_step(step=step, interval=nviz):
+            sim_visualization(
+                discr, eos, q, visualizer, vizname,
+                step=step, t=t,
+                exact_soln=exact_soln, viz_fields=viz_fields,
+                overwrite=overwrite, vis_timer=vis_timer
+            )
+    except SynchronizedError as err:
+        exception = err
+
+    terminate = True if exception is not None else False
+
+    if comm is None:
+        if terminate:
+            raise exception
+        return
+
+    from mpi4py import MPI
+
+    terminate = comm.allreduce(terminate, MPI.LOR)
+
+    if terminate:
+        # Log crash error message
+        if rank == 0:
+            logger.info(str(exception))
+            logger.info("Visualizing crashed state ...")
+        # Write out crashed field
+        sim_visualization(discr, eos, q,
+                          visualizer, vizname=vizname,
+                          step=step, t=t,
+                          viz_fields=viz_fields)
+        raise exception
```

Comment:
This too, since `exception` doesn't exist on all ranks.

This one is a little trickier to deal with I think... Seems like we can either:

1. Only re-raise the exception on the rank that originally raised it. I don't know if I like this since it makes any subsequent handling have to redo the MPI reduction.
2. Try to pickle the exception and broadcast it to all ranks. I vaguely remember trying this and running into issues, but unfortunately I can't recall the details.
3. Synchronize the error state before raising and then do the `raise` on all ranks.

Thoughts?

Reply:
@majosm I think option 3 (which I believe the code now sort of does) is probably the easiest for now? I don't want this PR to dig _too_ deeply into this. Are there any major objections to doing this: https://github.com/illinois-ceesd/mirgecom/pull/370/commits/4c938fa2cb2eef71d83aff54792fc14ad20b86f5?

Reply:
> To do option 3 properly I think we would need to synchronize before raising for each check inside `sim_healthcheck` (...)

Why not something along the lines of this that we talked about earlier?

```python
def fluid_healthcheck(...):
   nans_present = check_for_local_naninf()
   if nans_present:
      message += "NANs or INFs present in solution"
   health_problem = check_for_whatever_local_problem(...):
   if healt_problem:
      message += "Some other health issue."
   health = nansyes | healthproblem | yadda | yadda | yadda
   health = comm.allreduce(health, MPI.LOR)
   if health:
      log(message)
      toss TheException(message)
```

Reply:
> > To do option 3 properly I think we would need to synchronize before raising for each check inside `sim_healthcheck` (...)
> 
> Why not something along the lines of this that we talked about earlier?
> 
> ```python
> def fluid_healthcheck(...):
>    nans_present = check_for_local_naninf()
>    if nans_present:
>       message += "NANs or INFs present in solution"
>    health_problem = check_for_whatever_local_problem(...):
>    if healt_problem:
>       message += "Some other health issue."
>    health = nansyes | healthproblem | yadda | yadda | yadda
>    health = comm.allreduce(health, MPI.LOR)
>    if health:
>       log(message)
>       toss TheException(message)
> ```

Oops, yeah, I didn't mean to exclude this. I just meant that the error state needs to be synchronized before exceptions are raised.